### ISSUE TITLE
"New" Klaviyo integration: mapped fields are not submitting correctly

### ISSUE THREAD

**Issue by tremby** [verbb/formie#1986]
### Describe the bug

We are using the updated Klaviyo integration.

We noticed that queue jobs to run the integration are failing with errors like this:

```
API error: “{"errors":[{"id":"98c6a3bf-a155-435c-942b-3d39332b507a","status":400,"code":"invalid","title":"Invalid input.","detail":"'last_name' is not a valid field for the resource 'profile'.","source":{"pointer":"/data/attributes/profiles/data/0/attributes/last_name"},"links":{},"meta":{}}]}” /srv/cms-dev.my.website/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php:111
```

I tried removing the mapping it was complaining about and running the integration again, and then it complains about a different mapping. If I remove them all other than email address, it succeeds.

I've hit the "refresh" button and I believe the mappings it's trying to set up look correct.

The list we're trying to integrate with comes back like this from the server when I press that "refresh" button:

```json
        {
		"id": "SNC84W",
		"name": "Newsletter",
		"fields": [
			{
				"handle": "first_name",
				"name": "First Name",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "last_name",
				"name": "Last Name",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "email",
				"name": "Email",
				"type": null,
				"sourceType": null,
				"required": "1",
				"options": []
			},
			{
				"handle": "phone_number",
				"name": "Phone Number",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "city",
				"name": "City",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "region",
				"name": "Region",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "country",
				"name": "Country",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "zip",
				"name": "Zip",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "organization",
				"name": "Organization",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "title",
				"name": "Title",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "source",
				"name": "Source",
				"type": null,
				"sourceType": null,
				"required": null,
				"options": []
			},
			{
				"handle": "sms_consent",
				"name": "Consent To Send SMS",
				"type": "boolean",
				"sourceType": null,
				"required": null,
				"options": {
					"label": "Consent To Send SMS",
					"options": [
						{
							"label": "Yes",
							"value": true
						},
						{
							"label": "No",
							"value": false
						}
					]
				}
			}
		]
	}
```

As you can see, `last_name` certainly appears there.

### Steps to reproduce

1. Set up a "new" (as opposed to legacy) Klaviyo integration
2. Set up field mappings
3. Attempt to run the integration against a form submission

### Form settings

- Multi-page form: n/a
- Submission Method: n/a
- Client-side Validation: n/a
- Custom Form Templates: n/a

### Craft CMS version

4.10.6

### Plugin version

2.1.22

### Multi-site?

yes

**Comment by engram-design** [comment]
Fixed for the next release. To get this early, run `composer require verbb/formie:"dev-craft-4 as 2.1.22"`.

**Comment by tremby** [comment]
Thanks; I think this took us a step in the right direction.

Emitted earlier than the change you most recently made could affect, at Klaviyo.php line 161, when it tries to do the "profile-import" call, I'm getting the same error as before, but talking about the "zip" field.

I noticed in the docs at https://developers.klaviyo.com/en/reference/create_or_update_profile that the various location fields including `zip` are nested inside a `location` property in the request, whereas the structure returned by `getFieldMappingValues` and therefore being passed in as `attributes` is flat.

So in other words, being sent is something like

```json
{
  "first_name": "test",
  "last_name": "test",
  "email": "test@example.com",
  "zip": "a1a1a1"
}
```

while if I understand the docs right, it is expecting something like this

```json
{
  "first_name": "test",
  "last_name": "test",
  "email": "test@example.com",
  "location": {
    "zip": "a1a1a1"
  }
}
```

**Comment by engram-design** [comment]
Ah, my mistake on those location values, you're right. I've also added `address1` and `address2`.

To get this early, run `composer require verbb/formie:"dev-craft-4 as 2.1.22"`.

**Comment by tremby** [comment]
Thanks, I'll try that shortly.

Not strictly related, I think, but I've also noticed that validation on Klaviyo's end seems a lot stricter than I remember. It's enforcing a E.164 phone number format, for example, failing the queued integration job. And if I give it a phone number and email address, but it already has a record where that phone number is associated with a different email address, it'll error then too. This means people are going to submit their form and have no error, but then the queue job will fail shortly after, with no feedback to the user. I'm not sure if anything can really be done about that, any thoughts?

**Comment by engram-design** [comment]
I'll check out the phone number stuff shortly, that's good to know. As for duplicate handling, that's indeed a bit tricky when running it via a queue job, and I suppose depending on your use-case you might not always _want_ your user to know there's an issue with a duplicate email. For something like a newsletter signup, that's valid and common.

Other than disabling the queue altogether for Formie forms, there's not a lot of options for getting feedback for integrations to the user.

**Comment by tremby** [comment]
It seems to be happy now. Thanks for all your work.

**Comment by tremby** [comment]
We'd very much appreciate cutting a release sooner rather than later so we can have this and have the integration working, without pointing at a commit hash.

**Comment by engram-design** [comment]
Fixed in [2.1.23](https://github.com/verbb/formie/releases/tag/2.1.23)

**Comment by ccchapman** [comment]
> As for duplicate handling, that's indeed a bit tricky when running it via a queue job, and I suppose depending on your use-case you might not always _want_ your user to know there's an issue with a duplicate email. For something like a newsletter signup, that's valid and common.
> 
> Other than disabling the queue altogether for Formie forms, there's not a lot of options for getting feedback for integrations to the user.

@engram-design Would you entertain an "update or create" path for the Klaviyo integration? I don't know if that is something aligned with the expected behaviour of Formie marketing integrations. My recollection is that was the behaviour with the prior integration, though I could be mistaken.

**Comment by engram-design** [comment]
@ccchapman It's very possible there's different behaviour now due to the new API's, and we'll have to roll that handling ourselves, but that's pretty common. I might suggest tracking this in a new issue!
